flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java

```diff
@@ -482,8 +484,10 @@ protected void addElement(ReusableArrayData reused, E element) {
 
     @Override
     protected ArrayData buildList(ReusableArrayData list) {
-      list.setNumElements(writePos);
-      return list;
+      // Since ReusableArrayData is not accepted by Flink, use GenericArrayData temporarily to walk around it.
```

Comment:
Just realized the `ReusableArrayData` is actually from Iceberg code `FlinkParquetReaders`

Reply:
I don't really get it, i think the fix in Flink is included in 1.15.

Reply:
FLIP-27 source reader uses this util method to clone the RowData as it batches records for thread handover. It depends on `ArrayDataSerializer` to clone the field. With the change from PR #4712, `ReusableArrayData` object is not cloned and reused, which corrupts the batched RowData array.

```
  public static RowData clone(RowData from, RowData reuse, RowType rowType, TypeSerializer[] fieldSerializers) {
    GenericRowData ret;
    if (reuse instanceof GenericRowData) {
      ret = (GenericRowData) reuse;
    } else {
      ret = new GenericRowData(from.getArity());
    }
    ret.setRowKind(from.getRowKind());
    for (int i = 0; i < rowType.getFieldCount(); i++) {
      if (!from.isNullAt(i)) {
        RowData.FieldGetter getter = RowData.createFieldGetter(rowType.getTypeAt(i), i);
        ret.setField(i, fieldSerializers[i].copy(getter.getFieldOrNull(from)));
      } else {
        ret.setField(i, null);
      }
    }
    return ret;
  }
```

Reply:
@stevenzwu Do you have any failed test case? Let me have a look or reproduce it locally. I'm still confused why it still can not work.

Reply:
@yittg you can check out the dev branch for this PR and run the `TestIcebergSourceBounded`. `testCustomizedFlinkDataTypes` method should fail, because the array field has the same value for all records. 

For easier read of the diff, you can change the record count from 10 to 2.
```
List<Record> records = RandomGenericData.generate(schema, 10, 0L);
```

Reply:
Thanks @stevenzwu,
I finally got the point, the ArrayDataSerializer in Flink should be renew each time because it reuse the BinaryArrayData  internally. 'Think we can change the signature of RowDataUtil#clone to accept a supplier of serializer to walk around it for now.

Reply:
I looked into the Flink for why map data works but array data not with similar implementation, found that the MapDataSerializer#toBinaryMap always contains copy semantics implicitly but not for ArrayDataSerializer#toBinaryArray.